chain-impl-mockchain/src/testing/arbitrary/transaction.rs

```diff
@@ -0,0 +1,197 @@
+use crate::ledger::Ledger;
+use crate::testing::arbitrary::transaction::rand::prelude::SliceRandom;
+use crate::transaction::{Input, Output};
+use crate::utxo::{self, Entry};
+use crate::{
+    testing::address::{AddressData, AddressDataValue},
+    value::*,
+};
+use chain_addr::Address;
+use quickcheck::{Arbitrary, Gen};
+use std::iter;
+
+extern crate rand;
+use super::ArbitraryAddressDataVec;
+use rand::Rng;
+
+#[derive(Clone, Debug)]
+pub struct AribtraryValidTransactionData {
+    pub addresses: Vec<AddressDataValue>,
+    input_addresses: Vec<AddressDataValue>,
+    output_addresses: Vec<AddressDataValue>,
+}
+
+impl Arbitrary for AribtraryValidTransactionData {
+    fn arbitrary<G: Gen>(gen: &mut G) -> Self {
+        AribtraryValidTransactionDataBuilder::new(ArbitraryAddressDataVec::arbitrary(gen).0).build()
+    }
+}
+
+pub struct AribtraryValidTransactionDataBuilder {
+    source: Vec<AddressData>,
+}
+
+impl AribtraryValidTransactionDataBuilder {
+    pub fn new(source: Vec<AddressData>) -> Self {
+        AribtraryValidTransactionDataBuilder { source: source }
+    }
+
+    pub fn build(&self) -> AribtraryValidTransactionData {
+        let values: Vec<Value> = Self::generate_random_values()
+            .take(self.source.len())
+            .collect();
+
+        let addresses_values: Vec<AddressDataValue> =
+            Self::zip_addresses_and_values(&self.source, values);
+        let input_addresses_values = Self::choose_random_subset(&addresses_values);
+        let total_input_value = Self::sum_up_funds(
+            input_addresses_values
+                .iter()
+                .cloned()
+                .map(|x| x.value)
+                .collect(),
+        );
+        let output_addresses_values =
+            Self::choose_random_output_subset(&addresses_values, total_input_value);
+        AribtraryValidTransactionData::new(
+            addresses_values,
+            input_addresses_values,
+            output_addresses_values,
+        )
+    }
+
+    fn generate_random_values() -> impl Iterator<Item = Value> {
```

Comment:
Why do this and below functions need to be methods on `ArbitraryValidTransactionDataBuilder`? They don't depend on this type on any way.